### ISSUE TITLE
Permission denied while it Allowed! access to external storage problem

### ISSUE THREAD

**Issue by osamz** [burhanrashid52/PhotoEditor#226]
I tried the sample app on Android 10 (AVD virtual device), and it failed to saving photo on device, "permission denied", while I Allowed it.

I search and found that this deprecated function is the problem.
https://github.com/burhanrashid52/PhotoEditor/blob/e94dd4b595f8d3d965824a1492e45a04fee744d7/app/src/main/java/com/burhanrashid52/photoeditor/EditImageActivity.java#L266

I replace it with:
`File file = new File(this.getExternalFilesDir(DIRECTORY_PICTURES)
                    + File.separator + ""
                    + System.currentTimeMillis() + ".png");`
and it works fine, but it save photo in app private folder.

- I read that android 10 will not give direct access to external storage.

there are some workaround solutions, but I didn't use it.

**Comment by burhanrashid52** [comment]
Yes. I am aware of it. If would be really great if you can help to raise a PR for Android 10 storage support

**Comment by swentel** [comment]
FWIW, I was able to save it to the DCIM folder in a custom app I'm writing.

```
    String FOLDER_NAME = "Subfolder";
    File directory = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM), FOLDER_NAME);
    if (!directory.exists()) {
        exists = directory.mkdirs();
    }

    File file = null;
    if (exists) {
        file = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM)
                + File.separator + ""
                + FOLDER_NAME
                + File.separator + ""
                + System.currentTimeMillis() + ".png");
      }
```

I also added a media scan in the onSuccess method so the new file shows up immediately in the gallery apps

```
    MediaScannerConnection.scanFile(EditImageActivity.this, new String[] { finalFile.getAbsolutePath() }, null, null);
```

I then pass on the imageUri back to the calling activity where in onActivityResult I ask for permissions so the app can read/write the newly written file:

```
    final int takeFlags = data.getFlags()
        & (Intent.FLAG_GRANT_READ_URI_PERMISSION
        | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);

    getContentResolver().takePersistableUriPermission(uri, takeFlags);
```

Hope that might help a bit.

**Comment by ZaFaR97** [comment]
thanks, @swentel, @burhanrashid52

**Comment by breel93** [comment]
I just Opened PR about this Issue,
My fix:
`app:manifest.xml`
```xml
<application
        android:name="com.burhanrashid52.photoeditor.PhotoApp"
        ........
        android:requestLegacyExternalStorage="true">
```
`EditImageActivity`
Replace
```java
 File file = new File(Environment.getExternalStorageDirectory()
                    + File.separator + ""
                    + System.currentTimeMillis() + ".png");
```
with
```java
final File file = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES),
                System.currentTimeMillis() + ".png");
```
then add this method

```java
//added the photo file to gallery
    private static void galleryAddPic(Context context, String imagePath) {
        Intent mediaScanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
        File f = new File(imagePath);
        Uri contentUri = Uri.fromFile(f);
        mediaScanIntent.setData(contentUri);
        context.sendBroadcast(mediaScanIntent);
    }
```
call the method after in saved `onSuccess`
```java
    @Override
     public void onSuccess(@NonNull String imagePath) {
        ...........
        mPhotoEditorView.getSource().setImageURI(mSaveImageUri);
        galleryAddPic(EditImageActivity.this, file.getAbsolutePath());  <-- here
       }
```
I hope this helps someone

**Comment by nazmussakibsaad102** [comment]
@breel93 Thanks bro. Helped a lot :D

**Comment by Aldinsan1453** [comment]
Hi.. i'm still experiencing permission denied... while permission is already given in manifest... and also tried solution above but.. still permission is denied

**Comment by bjdev222** [comment]
check whet

> Hi.. i'm still experiencing permission denied... while permission is already given in manifest... and also tried solution above but.. still permission is denied

check whether u give permission to access storage from mobile setting for this app..

**Comment by burhanrashid52** [comment]
Fixed in #326
